hydromt/data_sources/geodataframe.py

```diff
@@ -0,0 +1,84 @@
+"""Generic DataSource for GeoDataFrames."""
+
+from logging import Logger
+from typing import Any, List, Optional
+
+import geopandas as gpd
+from pydantic import ValidationInfo, field_validator, model_validator
+
+from hydromt._typing import NoDataStrategy
+from hydromt.drivers.geodataframe_driver import GeoDataFrameDriver
+from hydromt.drivers.pyogrio_driver import PyogrioDriver
+
+from .data_source import DataSource
+
+# placeholder for proper plugin behaviour later on.
+_KNOWN_DRIVERS: dict[str, GeoDataFrameDriver] = {"pyogrio": PyogrioDriver}
+
+
+def driver_from_str(driver_str: str, **driver_kwargs) -> GeoDataFrameDriver:
+    """Construct GeoDataFrame driver."""
+    if driver_str not in _KNOWN_DRIVERS.keys():
+        raise ValueError(
+            f"driver {driver_str} not in known GeoDataFrameDrivers: {_KNOWN_DRIVERS.keys()}"
+        )
+
+    return _KNOWN_DRIVERS[driver_str](**driver_kwargs)
+
+
+class GeoDataFrameDataSource(DataSource):
+    """
+    DataSource for GeoDataFrames.
+
+    Reads and validates DataCatalog entries.
+    """
+
+    data_type = "GeoDataFrame"
+    driver: GeoDataFrameDriver
+
+    @model_validator(mode="before")
+    @classmethod
+    def _validate_data_type(cls, data: Any) -> Any:
+        if isinstance(data, dict):
+            if data.get("data_type", "") != "GeoDataFrame":
+                raise ValueError("'data_type' must be 'GeoDataFrame'.")
+        return data
```

Comment:
is this for internal sanity checks or does this doe anything for the user?

Reply:
Pydantic does not validate class variables, so I had to make a "before" validator to make sure that a `data_type` is included with the correct value. Otherwise the user can insert a wrong `data_type` and their mistake will be fixed silently.